File: day05/day05.rb
input = File.read('input.txt').strip.split('')

def react(data)
  stack = [data.first]
  data[1..-1].each { |x|
    if !stack.empty? and (stack.last.ord - x.ord).abs == 32
      stack.pop
    else
      stack << x
    end
  }
  stack.length
end
p react(input)

sizes = []
(?a..?z).each { |char|
  data = input.select { |x| x != char and x != char.upcase }
  sizes << react(data)
  }
p sizes.min

File: day05/day05.clj
(ns advent-2018.day-5
  (:require [clojure.string :as str]))

(def input (str/trim (slurp "input.txt")))

(defn reacts? [l r] (= 32 (Math/abs (- (int l) (int r)))))

(defn add-unit [res c]
  (if (some-> (peek res) (reacts c))
    (pop res) (conj res c)))

(defn react [in] (reduce add-unit [] in))

(count (react input))

(defn remove-units [x polymer]
  (remove (fn [y] (or (= x y) (reacts? x y))) polymer))

(->> (map char (range 97 123))
     (map #(count (react (remove-units % input))))
     (apply min))
